Add text formatting

InfusOnWoW opened the pull request:
TODO:
* multi selection
  * with different actions extra options don't appear at all
  * with different options, conditions are not highlighted as different

InfusOnWoW commented:
This still neeeds a bit of work, but is almost ready for some testing.

mrbuds commented:
"Class Colors as Player's GUID" require a GUID and does not work with %sourceName & %destName
Tooltip for the listing of available "special code" for combat log trigger does not specify %sourceGUID and %destGUID.
I don't know what's the best and/or most user friendly solution.

InfusOnWoW commented:
Right, the "true" solution is to make the state's variable more typed than they are currently. But that's a bigger change, so I kinda want to not do it atm. So the way I handle that for the class colored unit one is to add a description that explains it. Maybe that's good enough, if added to the guid variant?

I can add the guids to the %text, it was previously hidden because it wasn't useful. And now it would be useful.

mrbuds commented:
Mhmm maybe change option's name to
- Class Colors a Player's %GUID
- Class Colored %unit Name
with %GUID and %unit in red?

I agree with adding guids variable to combat log trigger display text tooltip, could do the same with %unit for aura trigger

mrbuds commented on WeakAuras/GenericTrigger.lua:
```diff
@@ -3887,6 +3904,40 @@ function GenericTrigger.GetTriggerDescription(data, triggernum, namestable)
 end
 
+do
+  -- Based on Code by DejaCharacterStats. Ugly code to figure out the
```
end of sentence missing

mrbuds commented on WeakAuras/GenericTrigger.lua:
```diff
@@ -3887,6 +3904,40 @@ function GenericTrigger.GetTriggerDescription(data, triggernum, namestable)
+      or class == "HUNTER" or class == "SHAMAN"
+      or class == "MAGE" or class == "PRIEST" or class == "WARLOCK"
+      or class == "DEATHKNIGHT" or class == "PALADIN" or class == "WARRIOR"
+    then
```
shouldn't the indentation match the "if" ?

mrbuds commented on WeakAuras/GenericTrigger.lua:
```diff
@@ -3887,6 +3904,40 @@ function GenericTrigger.GetTriggerDescription(data, triggernum, namestable)
+  elseif class == "MONK" then
+    function WeakAuras.CalculatedGcdDuration()
+      local spec = GetSpecialization() -- TODO check spec index
+      local primaryStat = select(6, GetSpecializationInfo(spec))
```
GetSpecializationInfo is nil on classic (nor does monks, so this comment is trash :D)

mrbuds commented on WeakAuras/Prototypes.lua:
```diff
@@ -6931,97 +7108,121 @@ end
+      elseif remaining > 60 then
+        remainingStr     = string.format("%i:", math.floor(remaining / 60));
+        remaining        = remaining % 60;
+        remainingStr     = remainingStr..string.format("%02i", remaining);
```
i would have written it like that (negligible i guess)
`remainingStr     = ("%s%02i"):format(remainingStr, remaining);`

InfusOnWoW replied:
I copied this code from the exstng formater.

mrbuds commented on WeakAuras/Prototypes.lua:
```diff
@@ -104,7 +104,9 @@ function WeakAuras.UnitChannelInfo(unit)
 
-
+local constants = {
+  nameRealmFilterDesc = L["Filter formats: 'Name', 'Name-Realm', '-Realm'. Multiple Filters can be separated by ','."],
```
For multiple entries it could use the sentence as the tooltip for encounter ids , zones, etc...
"\n\nSupport multiple entries, separated by commas"

mrbuds commented on WeakAurasOptions/BuffTrigger2.lua:
```diff
@@ -743,6 +740,38 @@ local function GetBuffTriggerOptions(data, optionTriggerChoices)
+      type = "input",
+      width = WeakAuras.normalWidth,
+      name = L["Unit Name Filter"],
+      desc = L["Filter formats: 'Name', 'Name-Realm', '-Realm'. Multiple Filters can be separated by ','."],
```
Same here

mrbuds commented on WeakAuras/BuffTrigger2.lua:
```diff
@@ -2403,6 +2413,7 @@ function BuffTrigger.GetAdditionalProperties(data, triggernum)
   ret = ret .. "|cFFFF0000%".. triggernum .. ".debuffClassIcon|r - " .. L["Debuff Class Icon"] .. "\n"
   ret = ret .. "|cFFFF0000%".. triggernum .. ".unitCaster|r - " .. L["Caster Unit"] .. "\n"
   ret = ret .. "|cFFFF0000%".. triggernum .. ".casterName|r - " .. L["Caster Name"] .. "\n"
+  ret = ret .. "|cFFFF0000%".. triggernum .. ".unit|r - " .. L["Unit"] .. "\n"
```
can you also add sourceGUID and destGUID for combat log trigger text tooltip ?

mrbuds commented:
```
1x WeakAurasOptions\Cache.lua:45: attempt to call global 'GetCategoryList' (a nil value)
WeakAurasOptions\Cache.lua:45: in function <WeakAurasOptions\Cache.lua:23>
```
on classic

nullKomplex commented:
Some inconsistencies:

Setting it to Class Colored %unit Name works on units in the group, but on anything not in the group it's blank. I would expect it to simply be white instead. Setting it to Class Colors a Player's %guid, Time in GCDs, causes an error on closing WA options that prevents auras from resuming (and bricks the addon on reload). Time Format leaves it blank as well but I expected that one. AbbreviateLargeNumbers and AbbreviateNumbers did not leave it blank however.

I'm not sure how this happens, but I did `%destName %p` and I get the GUID error with no formatting options.

```3x WeakAuras\Types.lua:105: Usage: GetPlayerInfoByGUID("playerGUID")
[string "=[C]"]: in function `GetPlayerInfoByGUID'
[string "@WeakAuras\Types.lua"]:105: in function `formatter'
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6542: in function <WeakAuras\WeakAuras.lua:6520>
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6691: in function <WeakAuras\WeakAuras.lua:6661>
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6772: in function `ReplacePlaceHolders'
[string "@WeakAuras\SubRegionTypes\SubText.lua"]:256: in function `?'
[string "@WeakAuras\RegionTypes\RegionPrototype.lua"]:33: in function `Notify'
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6292: in function <WeakAuras\WeakAuras.lua:6289>
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6375: in function <WeakAuras\WeakAuras.lua:6342>
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6459: in function `UpdatedTriggerState'
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6229: in function `UpdateFakeStatesFor'
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6202: in function `FakeStatesFor'
...
[string "@CorruptionTooltips\Libs\AceConfig-3.0-3\AceConfigDialog-3.0\AceConfigDialog-3.0-79.lua"]:51: in function <...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0.lua:49>
[string "@CorruptionTooltips\Libs\AceConfig-3.0-3\AceConfigDialog-3.0\AceConfigDialog-3.0-79.lua"]:843: in function <...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0.lua:664>
[string "=[C]"]: ?
[string "@CorruptionTooltips\Libs\AceGUI-3.0\AceGUI-3.0-41.lua"]:72: in function <...ns\CorruptionTooltips\Libs\AceGUI-3.0\AceGUI-3.0.lua:70>
[string "@CorruptionTooltips\Libs\AceGUI-3.0\AceGUI-3.0-41.lua"]:306: in function `Fire'
[string "@CorruptionTooltips\Libs\AceGUI-3.0-41\widgets\AceGUIWidget-DropDown.lua"]:442: in function <...ps\Libs\AceGUI-3.0\widgets\AceGUIWidget-DropDown.lua:433>
[string "=[C]"]: ?
[string "@CorruptionTooltips\Libs\AceGUI-3.0\AceGUI-3.0-41.lua"]:72: in function <...ns\CorruptionTooltips\Libs\AceGUI-3.0\AceGUI-3.0.lua:70>
[string "@CorruptionTooltips\Libs\AceGUI-3.0\AceGUI-3.0-41.lua"]:306: in function `Fire'
[string "@TomTom\libs\AceGUI-3.0-41\widgets\AceGUIWidget-DropDown-Items.lua"]:351: in function <...s\AceGUI-3.0\widgets\AceGUIWidget-DropDown-Items.lua:341>
```

```2x WeakAuras\Types.lua:303: attempt to compare string with number
[string "@WeakAuras\Types.lua"]:303: in function `formatter'
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6542: in function <WeakAuras\WeakAuras.lua:6520>
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6691: in function <WeakAuras\WeakAuras.lua:6661>
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6772: in function `ReplacePlaceHolders'
[string "@WeakAuras\SubRegionTypes\SubText.lua"]:256: in function `?'
[string "@WeakAuras\RegionTypes\RegionPrototype.lua"]:33: in function `Notify'
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6292: in function <WeakAuras\WeakAuras.lua:6289>
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6375: in function <WeakAuras\WeakAuras.lua:6342>
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6459: in function `UpdatedTriggerState'
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6229: in function `UpdateFakeStatesFor'
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:6202: in function `FakeStatesFor'
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:4705: in function <WeakAuras\WeakAuras.lua:4593>
...
[string "@CorruptionTooltips\Libs\AceConfig-3.0-3\AceConfigDialog-3.0\AceConfigDialog-3.0-79.lua"]:51: in function <...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0.lua:49>
[string "@CorruptionTooltips\Libs\AceConfig-3.0-3\AceConfigDialog-3.0\AceConfigDialog-3.0-79.lua"]:843: in function <...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0.lua:664>
[string "=[C]"]: ?
[string "@CorruptionTooltips\Libs\AceGUI-3.0\AceGUI-3.0-41.lua"]:72: in function <...ns\CorruptionTooltips\Libs\AceGUI-3.0\AceGUI-3.0.lua:70>
[string "@CorruptionTooltips\Libs\AceGUI-3.0\AceGUI-3.0-41.lua"]:306: in function `Fire'
[string "@CorruptionTooltips\Libs\AceGUI-3.0-41\widgets\AceGUIWidget-DropDown.lua"]:442: in function <...ps\Libs\AceGUI-3.0\widgets\AceGUIWidget-DropDown.lua:433>
[string "=[C]"]: ?
[string "@CorruptionTooltips\Libs\AceGUI-3.0\AceGUI-3.0-41.lua"]:72: in function <...ns\CorruptionTooltips\Libs\AceGUI-3.0\AceGUI-3.0.lua:70>
[string "@CorruptionTooltips\Libs\AceGUI-3.0\AceGUI-3.0-41.lua"]:306: in function `Fire'
[string "@TomTom\libs\AceGUI-3.0-41\widgets\AceGUIWidget-DropDown-Items.lua"]:351: in function <...s\AceGUI-3.0\widgets\AceGUIWidget-DropDown-Items.lua:341>```

InfusOnWoW commented:
@nullKomplex you need to include auras that show the problems, from your minimal descriptions I cannot figure out what the setup is.

nullKomplex commented:
Well the weird `%destName %p` was happening with any aura I tried it on, but now I can't recreate it, even with the exact auras. The others happen with _literally any_ aura.

InfusOnWoW commented:
I have no idea what the trigger for "works on units in the group, but on anything not in the group it's blank" is, and no that is not independant of the trigger.

nullKomplex commented:
Everything I described is re-creatable with that aura by **only** changing the formatting option and sending any chat message. Oddly enough, it returns an empty string even for myself when I use it for that trigger, maybe because it includes server even for myself.

mrbuds commented:
If i select the format "Class Colors a PLayer's %guid" for %s of an aura trigger (which obviously won't work) i get this error
```
2x WeakAuras\Types.lua:105: Usage: GetPlayerInfoByGUID("playerGUID")
[C]: in function `GetPlayerInfoByGUID'
WeakAuras\Types.lua:105: in function `formatter'
WeakAuras\WeakAuras-2.17.5-26-g1a22010.lua:6542: in function <WeakAuras\WeakAuras.lua:6520>
WeakAuras\WeakAuras-2.17.5-26-g1a22010.lua:6712: in function `ReplacePlaceHolders'
WeakAuras\SubRegionTypes\SubText.lua:256: in function `?'
```
guid could be tested with guid:match("Player%-%w+%-%w+")

InfusOnWoW commented:
@nullKomplex, @mrbuds I've fixed the bugs you've found. Thanks to your testing. 

Also apparently the naming of the "class colored unit/guid" option was confussing, at least I think null didn't understand how they work. They don't format a %sourceName, but a %sourceUnit.

To make that better, I've renamed those two options and slightly changed how they look like. That also adds a good place for e.g. limiting character names to x chars.

Also, I think there are two open questions, that I'm very much looking for feedback:

a) The selection of text formatters. Which ones do we want to provide?
b) Is that gcd idea a good idea at all? I think if I'm the only one that finds that idea interesting, then I'll drop that again.

nullKomplex commented on WeakAuras/Types.lua:
```diff
@@ -66,15 +66,273 @@ WeakAuras.group_hybrid_sort_types = {
+    end
+  },
+  Unit = {
+    display = L["Fromats |cFFFF0000%unit|r"],
```
Fromats

InfusOnWoW replied:
Thanks, fixed.

nullKomplex commented:
No, I was testing for inconsistencies regarding data in and out, like I said. It's inconsistent that most of them, when given input that does not match the formatter (e.g. a string to Round), return the input, while others (like class colored) return a blank string. I would prefer it (and honestly expect it to be DoA if not) if class coloring a unit, which works on group members as they are valid unit IDs, would still return the string you specified (say, a pet or an enemy) if the unit cannot be colored.

Also was Time Format the default before? I notice with that with negative values None/Round now show that negative value for progress, whereas before I was seeing empty strings.

InfusOnWoW commented:
DoA?

InfusOnWoW commented:
I think I can make the unit formatting return the un modified string if it's not a unit.

Though it's quite melodramic to pronounce a formatter "dead" because it does not support names.

InfusOnWoW commented:
I think I should fix Round + timed to not show negative values.

InfusOnWoW commented:
<img width="560" alt="WowClassic_Glv74Q9xol" src="https://user-images.githubusercontent.com/12950521/85926877-af17b280-b8a2-11ea-981c-a109242f5a28.png">
<img width="552" alt="WowClassic_gU9bI5dF0V" src="https://user-images.githubusercontent.com/12950521/85926880-b212a300-b8a2-11ea-8407-1b45d14f5ae7.png">
<img width="566" alt="WowClassic_kcugphC54c" src="https://user-images.githubusercontent.com/12950521/85926888-b50d9380-b8a2-11ea-82a5-6bf607809a83.png">
<img width="548" alt="iN1Nj0uJRb" src="https://user-images.githubusercontent.com/12950521/85926894-b76fed80-b8a2-11ea-8673-e23c32918aea.png">
<img width="254" alt="mXH1hxjOnR" src="https://user-images.githubusercontent.com/12950521/85926896-b9d24780-b8a2-11ea-962c-e3df0c731918.png">
<img width="546" alt="WowClassic_q2qVnZ3cMB" src="https://user-images.githubusercontent.com/12950521/85926899-bccd3800-b8a2-11ea-9bbf-05797616daf0.png">
<img width="552" alt="4ArjV17CBr" src="https://user-images.githubusercontent.com/12950521/85926901-be96fb80-b8a2-11ea-874d-a373b39394d8.png">

Stanzilla commented:
While I like having more options this all feels like overengineering to the max :D

InfusOnWoW commented:
I'm actually planing on extending the time formatting options after this PR is in....

I actually think the options are mostly the ones that come up regularly on the discord, except of the gcd option.
